### PR TITLE
added the Avi ansible lookup module

### PR DESCRIPTION
**Pull request by ericsysmin**
##### SUMMARY
Adding new lookup plugin to allow lookups against Avi controller to get values

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
avi

##### ADDITIONAL INFORMATION

```paste below
- debug: msg="{{ lookup('avi', obj_name='vs1', obj_type='virtualservice',
    tenant='admin', controller_ip='controller_ip', username='username', password='password') }}"
- debug: msg="{{ lookup('avi', obj_uuid='virtualservice-5c0e183a-690a-45d8-8d6f-88c30a52550d',
    obj_type='virtualservice', tenant='admin', controller_ip='controller_ip', username='username',
    password='password') }}"
```

### CODE REVIEW

File: lib/ansible/plugins/lookup/avi.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        try:
+            session_params = {
+                'controller_ip': kwargs.get('controller_ip', None),
+                'username': kwargs.get('username', None),
+                'password': kwargs.get('password', None),
+                'token': kwargs.get('token', None),
+                'tenant': kwargs.get('tenant', None),
+                'tenant_uuid': kwargs.get('tenant_uuid', None),
+                'verify': kwargs.get('verify', None),
+                'port': kwargs.get('port', None),
```
Comment (grastogi23) [inline lib/ansible/plugins/lookup/avi.py:58]: I would suggest we don't expose everything. let us retrict it to the options allowed in the ansible modules.

File: lib/ansible/plugins/lookup/avi.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            avi = ApiSession(**session_params)
+            if kwargs.get('obj_name', None):
+                path = kwargs['obj_type']
+                name = kwargs['obj_name']
+                response = avi.get_object_by_name(path, name, **get_params)
+            else:
+                obj_type = kwargs['obj_type']
+                obj_uuid = kwargs['obj_uuid']
+                path = "%s/%s" % (obj_type, obj_uuid)
+                response = avi.get(path, **get_params).json()
```
Comment (grastogi23) [inline lib/ansible/plugins/lookup/avi.py:85]: The format of the return would be different. So, let us discuss what is recommended in ansible.

**Review by grastogi23 — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/lookup/avi.py (hunk 1 above)
Comment (ericsysmin) [reply lib/ansible/plugins/lookup/avi.py:58]: The problem was is how they are submitted via lookup, there's no way to seperate them unless we unlike every other lookup plugin do a dict only var for auth. Our session function does not support kwargs

File: lib/ansible/plugins/lookup/avi.py (hunk 2 above)
Comment (ericsysmin) [reply lib/ansible/plugins/lookup/avi.py:85]: What do you mean format of the return? It returns a proper dict and works on lookup, as tested. the get_object_by_name actually already returns the json object, get does not.

File: lib/ansible/plugins/lookup/avi.py (hunk 1 above)
Comment (ericsysmin) [reply lib/ansible/plugins/lookup/avi.py:58]: Another option is to rewrite the module_utils.avi to include ability to process requests from plugins and not from modules...but that's also why Google uses a MockupModule class to emulate a module value for sessions.

```
class GcpMockModule(object):
    def __init__(self, params):
        self.params = params

    def fail_json(self, *args, **kwargs):
        raise AnsibleError(kwargs['msg'])

    def raise_for_status(self, response):
        try:
            response.raise_for_status()
        except getattr(requests.exceptions, 'RequestException'):
            self.fail_json(msg="GCP returned error: %s" % response.json())


class GcpFileLookup():
    def get_file_contents(self, module):
        auth = GcpSession(module, 'storage')```

**Comment by grastogi23** [comment]
Based on the Ansible documentation lookup should return list and it should take the input as filter
https://docs.ansible.com/ansible/latest/dev_guide/developing_plugins.html#lookup-plugins

In that case we should not restrict it to a single object return. This code can be very powerful to allow it to return list of avi objects as the standard interface and inline with lookup plugins.

**Comment by grastogi23** [comment]
please use the aws as the reference
https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/lookup/aws_account_attribute.py

**Comment by chaitanyaavi** [comment]
Create another by Sandeep so closing this PR

**Pull request closed without merging**